src/main/java/com/epam/javacourses2016/task18/SolverTask18.java

```diff
@@ -10,24 +12,100 @@
+            } else {
+                pairsMatrix.get(i).set(matrix.getWidth() - 1, new Pair(1,1));
+            }
+        }
+        int maxS = 1;
+        int maxI = 0;
+        int maxJ = 0;
+        for (int i = matrix.getHeight() - 2; i >= 0; i--) {
+            for (int j = matrix.getWidth() - 2; j >= 0; j--) {
+                if (matrix.getValue(i, j) == matrix.getValue(i, j + 1) && matrix.getValue(i, j) != matrix.getValue(i + 1, j)) {
+                    Pair temp = new Pair(1, pairsMatrix.get(i+1).get(j).getWidth() + 1);
+                    if (temp.getWidth() * temp.getHeight() >= maxS) {
+                        maxS = temp.getWidth() * temp.getHeight();
+                        maxI = i;
+                        maxJ = j;
+                    }
+                    pairsMatrix.get(i).set(j, temp);
+                } else if (matrix.getValue(i, j) != matrix.getValue(i, j + 1) && matrix.getValue(i, j) == matrix.getValue(i + 1, j)) {
+                    Pair temp = new Pair(pairsMatrix.get(i).get(j+1).getHeight() + 1, 1);
+                    pairsMatrix.get(i).set(j, temp);
+                    if (temp.getWidth() * temp.getHeight() >= maxS) {
+                        maxS = temp.getWidth() * temp.getHeight();
+                        maxI = i;
+                        maxJ = j;
+                    }
+                } else if (matrix.getValue(i, j) == matrix.getValue(i, j + 1) && matrix.getValue(i, j) == matrix.getValue(i + 1, j)) {
+                    Pair temp = new Pair(pairsMatrix.get(i+1).get(j).getHeight() + 1, pairsMatrix.get(i).get(j+1).getWidth() + 1);
+                    Pair temp1 = new Pair(Math.min(pairsMatrix.get(i).get(j+1).getHeight() + 1, pairsMatrix.get(i+1).get(j).getHeight()),
+                            Math.min(pairsMatrix.get(i+1).get(j).getWidth(), pairsMatrix.get(i).get(j+1).getWidth()));
+                   if (temp.getHeight() * temp.getWidth() > temp1.getHeight() * temp1.getWidth()) {
+                       pairsMatrix.get(i).set(j, temp);
+                   } else {
+                       pairsMatrix.get(i).set(j, temp1);
+                   }
+                    if (temp.getWidth() * temp.getHeight() >= maxS) {
+                        maxS = temp.getWidth() * temp.getHeight();
+                        maxI = i;
+                        maxJ = j;
+                    }
+                } else {
+                    pairsMatrix.get(i).set(j, new Pair(1,1));
+                }
+            }
+        }
+        int resMatrixHeight = pairsMatrix.get(maxI).get(maxJ).getHeight();
+        int resMatrixWidth = pairsMatrix.get(maxI).get(maxJ).getWidth();
+        int[][] resM = new int[resMatrixHeight][resMatrixWidth];
+        for (int i = maxI; i < maxI + resMatrixHeight; i++) {
+            for (int j = maxJ; j < maxJ + resMatrixWidth; j++) {
+                        resM[i - maxI][j - maxJ] = matrix.getValue(i,j);
+            }
+        }
+        return new Matrix(resM);
```

Comment:
Your code is not in Java style. More like C style.
You should decompose big methods, exclude copy-pasting. Have a look at https://www.ozon.ru/context/detail/id/5011068/, https://www.ozon.ru/context/detail/id/1308678/